crates/accelerate/src/euler_one_qubit_decomposer.rs

```diff
@@ -571,21 +577,116 @@ pub fn generate_circuit(
     Ok(res)
 }
 
-#[derive(Clone, Debug, Copy)]
+const EULER_BASIS_SIZE: usize = 12;
+
+static EULER_BASES: [&[&str]; EULER_BASIS_SIZE] = [
+    &["u3"],
+    &["u3", "u2", "u1"],
+    &["u"],
+    &["p", "sx"],
+    &["u1", "rx"],
+    &["r"],
+    &["rz", "ry"],
+    &["rz", "rx"],
+    &["rz", "rx"],
+    &["rx", "ry"],
+    &["rz", "sx", "x"],
+    &["rz", "sx"],
+];
+static EULER_BASIS_NAMES: [EulerBasis; EULER_BASIS_SIZE] = [
+    EulerBasis::U3,
+    EulerBasis::U321,
+    EulerBasis::U,
+    EulerBasis::PSX,
+    EulerBasis::U1X,
+    EulerBasis::RR,
+    EulerBasis::ZYZ,
+    EulerBasis::ZXZ,
+    EulerBasis::XZX,
+    EulerBasis::XYX,
+    EulerBasis::ZSXX,
+    EulerBasis::ZSX,
+];
+
+/// A structure containing a set of supported `EulerBasis` for running 1q synthesis
+#[derive(Debug, Clone)]
+pub struct EulerBasisSet {
+    basis: [bool; EULER_BASIS_SIZE],
+    initialized: bool,
+}
+
+impl EulerBasisSet {
+    // Instantiate a new EulerBasisSet
+    pub fn new() -> Self {
+        EulerBasisSet {
+            basis: [false; EULER_BASIS_SIZE],
+            initialized: false,
+        }
+    }
+
+    /// Return true if this has been initialized any basis is supported
+    pub fn initialized(&self) -> bool {
+        self.initialized
+    }
+
+    /// Add a basis to the set
+    pub fn add_basis(&mut self, basis: EulerBasis) {
+        self.basis[basis as usize] = true;
+        self.initialized = true;
+    }
+
+    /// Get an iterator of all the supported EulerBasis
+    pub fn get_bases(&self) -> impl Iterator<Item = EulerBasis> + '_ {
+        self.basis
+            .iter()
+            .enumerate()
+            .filter_map(|(index, supported)| {
+                if *supported {
+                    Some(EULER_BASIS_NAMES[index])
+                } else {
+                    None
+                }
+            })
+    }
+
+    /// Check if a basis is supported by this set
+    pub fn basis_supported(&self, basis: EulerBasis) -> bool {
+        self.basis[basis as usize]
+    }
+
+    /// Modify this set to support all EulerBasis
+    pub fn support_all(&mut self) {
+        self.basis = [true; 12];
+        self.initialized = true;
+    }
+
+    /// Remove an basis from the set
+    pub fn remove(&mut self, basis: EulerBasis) {
+        self.basis[basis as usize] = false;
+    }
+}
+
+impl Default for EulerBasisSet {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Clone, Debug, Copy, Eq, Hash, PartialEq)]
 #[pyclass(module = "qiskit._accelerate.euler_one_qubit_decomposer")]
 pub enum EulerBasis {
-    U321,
-    U3,
-    U,
-    PSX,
-    ZSX,
-    ZSXX,
-    U1X,
-    RR,
-    ZYZ,
-    ZXZ,
-    XYX,
-    XZX,
+    U3 = 0,
+    U321 = 1,
+    U = 2,
+    PSX = 3,
+    U1X = 4,
+    RR = 5,
+    ZYZ = 6,
+    ZXZ = 7,
+    XZX = 8,
+    XYX = 9,
+    ZSXX = 10,
+    ZSX = 11,
```

Comment:
What are these indices used for? I'm sure it's there but I haven't been able to find it in the code

Reply:
It's to force an specific integer for each variant, to be explicit so that the variant matches it's position in each of the static arrays. It's implicit with the ordering, but it's best to be explicit to avoid potential issues since we're doing all the look ups in the statics.